[28기 강희원][Refactor] Review 반영 진행중

KH2Wone opened the pull request:
## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [x] 기능 추가
- [x] 리뷰 반영
- [x] 리팩토링
- [x] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
- 리액트 연결
- [x] 개인 폴더(pages/heewonkang)에 개인이 작업하던 Login, Main 코드 옮기기
- React 과제
- [x] Mission 1 : 사용자 입력 데이터 저장
- [x] Mission 2 : 로그인 버튼 활성화
- [x] Mission 3 : 댓글 기능
- [x] Mission 4 : 댓글 컴포넌트화 + props로 데이터 전달
- [x] Mission 5 : map 함수 적용시 key props를 부여하는 이유
- [x] Mock Data 작업
- 리팩토링

<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)
1. 페이지 이동
`element="/login-won"`
`element="/Main-won"`
2. 로그인
- 유효성 검사 (아이디 '@' 포함 && 비밀번호 5글자 이상)
- 로그인 버튼 활성화
- 아이디 및 비밀번호 입력시에만 메인으로 넘어갈 수 있도록 로직 구현 (진행중)
3. 메인
- 댓글 기능
- 댓글 컴포넌트화

<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)
- 협업의 첫 과정을 천천히 따라갈 수 있었습니다.
- setState를 활용하여 값을 조건에 따라 동적으로 변경될 수 있도록 이벤트를 주었습니다.
- sider 문제 해결: setState의 값을 사용하지 않으면 sider에 걸려서 코드를 모두 작성한 후 커밋해주었습니다.
- 댓글 기능에 버그가 있어서 하나하나 console을 찍어보면서 찾았더니 state값으로 제어한게 아니라 받아온 데이터로 제어를 해서 버그가 생긴거였습니다. state의 개념을 확고히 하는데 큰 공부가 되었습니다.


<br />

## :: 기타 질문 및 특이 사항

KH2Wone committed:
Add: Login, Main 파일 생성

KH2Wone committed:
Update: .scss 변수 호출하여 적용

KH2Wone committed:
Add : Mission 1 - 사용자 입력 데이터 저장 기능 구현

KH2Wone committed:
Fix : side error - never used.

KH2Wone committed:
Add : Mission 2 - 로그인 버튼 활성화 기능 구현

KH2Wone committed:
[Add] State, Props, Event 과제 |  Mission 3, 4 완료
- Add : Mission 3 - 댓글 기능 구현
- Add : Mission 4 - 댓글 컴포넌트화, props로 데이터 전달 구현

KH2Wone committed:
Refactor: 불필요한 내용 제거 및 구조분해할당 적용

seankim1111 reviewed (commented):
고생하셨습니다 희원님! 코드가 깔끔하네요🙂

YunKukPark commented on src/pages/heewonKang/Login/FormLogin.js:
```diff
@@ -0,0 +1,73 @@
+    }
+  };
+
+  const handleIdInput = event => {
+    const { value } = event.target;
+    setId(value);
+    const loginBtn = document.querySelector('#login_btn');
+    if (value.includes('@') && pwValue.length > 5) {
+      loginBtn.classList.add('active_login');
+    } else {
+      loginBtn.classList.remove('active_login');
+    }
+  };
+
+  const handlePwInput = event => {
+    const { value } = event.target;
+    setPw(value);
+    const loginBtn = document.querySelector('#login_btn');
+    if (value.length > 5 && idValue.includes('@')) {
+      loginBtn.classList.add('active_login');
+    } else {
+      loginBtn.classList.remove('active_login');
+    }
+  };
```
무언가 같은 행동을 하는 것 같아서 마구마구 묶어주면 좋을 것 같습니다 🤔

하나의 handleInput을 만들어 놓고 들어오는 값의 id에 따라 다른 행동을 취하게 해주면 어떨까요?? 
handleId 와 handlePw 는 모두 
1. event.target.value 를 `state`에 셋팅해주고 
2. Validate 를 하고,
3. 로그인 버튼을 변화 시켜주고 있는 것 같습니다

아래는 예시입니다! 
```jsx
const handleInput = event => {
    const {
      target: { id, value },  // event.target.id 와 value를 받아옵니다.
    } = event;
    if (id === 'login_id') setId(value);
    if (id === 'login_pw') setPw(value);
    
    // Do Validate (state에 set된 내용을 가지고)
    // Do Handle Login Button (validate 된 내용을 가지고)
  };
```

KimJeongHyun commented on src/pages/heewonKang/Main/Main.js:
```diff
@@ -1,3 +1,343 @@
+                  </div>
+                  <ul className="story">
+                    <li>
+                      <div>
```
혹시 위와 같이 div를 하나하나 묶어준 이유가 있을까요???

KH2Wone replied:
좌우 스크롤이 가능하도록 만드느라 껍데기를 많이 씌웠는데요, 정현님이 말씀해주시고 다시 코드확인을 해보니까 리팩토링을 해봐야겠네요!! 언급해주셔서 감사합니다 💡

KimJeongHyun commented on src/pages/heewonKang/Main/Main.js:
```diff
@@ -1,3 +1,343 @@
+
+  return (
+    <div id="container">
+      <NavWon />
```
Navigation bar빼신 것 처럼 몇몇 부분 컴포넌트로 빼시면 좋을 것 같습니다!

KimJeongHyun commented on src/pages/heewonKang/Main/Main.js:
```diff
@@ -1,3 +1,343 @@
+                </ul>
+                <ul className="feed_img_dots">
+                  <li />
+                  <li className="active_feed_dots" />
```
저와 마찬가지로 레이아웃 설정용으로 static 하게 값을 설정한 것 같아요!
약간 저도 기억하고자 메모를 남긴다면

나중에 BE와 연동해서 데이터를 가져온다고 가정했을 때

여기 feed_dots도 map 함수로 돌리고, key에 이를테면 imgID 같은 걸 넘겨주고
현재 보고 있는 이미지의 ID에 따라서 각각의 dot을 state의 변화에 따라 CSS를 업데이트해주는 방법으로 해주면
좋을 것 같습니다!

KH2Wone replied:
오오 그렇게 하는게 좋겠네요 감사합니다 ‼️

KimJeongHyun commented on src/pages/heewonKang/Main/Main.js:
```diff
@@ -1,3 +1,343 @@
+                    <span>hello world 🌏</span>
+                  </div>
+                  <a href="#!" className="more_comment">
+                    댓글 <b>120개</b> 모두 보기
```
피드도 하나하나 들어오면 map으로 돌리고 각각의 피드는 각기 다른 댓글 갯수와 댓글들을 가지게 수정하게 되면
 {comment.length}를 통해 동적으로 대처하면 좋을 것 같습니다!

KH2Wone replied:
넵!! 피드는 컴포넌트로 분리하고 map 함수도 돌렸습니다!! 훨씬 낫네요 ㅎㅎ
각각의 피드는 각기 다른 댓글 갯수,,,, 각기 다른 댓글 ,,, 이건 어렵게 느껴지는데 한번 해볼게요. 감사합니다 !!!

YunKukPark commented on src/pages/heewonKang/Main/Main.js:
```diff
@@ -1,3 +1,343 @@
+  };
+
+  const addComment = () => {
+    setComment(current => {
+      const newList = [...current];
+      const newComment = {
+        id: newList.length + 1,
+        name: 'kimCode',
+        comment: inputComment,
+      };
+      newList.push(newComment);
+      setInputComment('');
+      return newList;
+    });
```
불변성을 위해 deepCopy를 하여 사용 하셨군요! 
혹시 전개연산자를 이렇게 사용해보시는 건 어떨까요??

```jsx
const newComment = { 어쩌구 저쩌구 }; 
setComment([...comment, newComment]); //  바로 넣어줄때 한번 전개연산 후 뒤에 newComment를 붙혀줬습니다.
```

KH2Wone replied:
윤국님...... 크으으으으으 👍🏻
저 진짜 코드 바꾸고 신세계 경험했습니다 ㅋㅋㅋ 저 진짜 벅차올라요(?) ㅋㅋㅋㅋㅋ 이렇게 쓸 수 있다니... 세상에...
덕분에 리팩토링 제대로 했습니다, 처음에 윤국님이 추천해주신 방식을 봤을 때는 '왜 이렇게 해야할까? 어떤 이유때문에 추천해주셨을까?', '어렵다' 이렇게 느껴졌는데 직접 하나하나 작성하면서 이해하다보니 감탄밖에 안나옵니다 ... 감사해요!!

YunKukPark commented on src/pages/heewonKang/Main/Main.js:
```diff
@@ -1,3 +1,343 @@
+                    댓글 <b>120개</b> 모두 보기
+                  </a>
+                  <ul className="comment_list">
+                    {comment.map((item, index) => {
+                      return <Comment key={index} commentItem={item} />;
+                    })}
```
이렇게 줄여써도 좋을 것 같아요 ~

```jsx
{ comment.map((item, index) => <Comment key={index} commentItem={item} />) }
```

한 줄에 쓸 수 있는 일이면 중괄호와 return 없이 !

KH2Wone replied:
아 그렇네요!! 어차피 return해주는 값이니까요, 체크해주셔서 감사합니다 🔥

YunKukPark commented on src/pages/heewonKang/Login/Login.scss:
```diff
@@ -0,0 +1,155 @@
+            text-align: center;
+          }
+
+          .login_form {
+            width: 100%;
+            margin: 36px 0px 10px;
+
+            form {
+              ul {
+                li {
+                  width: 100%;
+
+                  &:not(li.login_btn) {
+                    margin: 0 auto 6px;
+                    border: 1px solid rgba(var(--b6a, 219, 219, 219), 1);
+                    border-radius: 3px;
+                    display: flex;
+                    height: 36px;
+                  }
```
전체적으로 필요에 따라 사용하신 & 기호나 :not 등으로 스타일링 문서를 깔끔하게 사용하시는 것 같아요 🙌

KH2Wone committed:
Refactor : useEffect를 활용해서 리팩토링

KH2Wone committed:
[Refactor] component 분리 및 로그인 버튼 로직 변경
- component 분리: Aside, Feed, InfoList, UsersStory
- 로그인 버튼 로직 변경: 로그인 버튼 활성화 되면 바로 메인으로 이동 되도록 변경

KH2Wone committed:
Add : 백엔드와 통신

KH2Wone committed:
[Refactor] feed, 댓글 기능 리팩토링
- 댓글 기능: 윤국님 코드리뷰 내용으로 리팩토링 (불필요한 return, setComment에서 전개 연산자 사용하여 가독성을 높임)
- 피드 여러개 관리: map 함수 사용하여 각각의 피드 데이터 받아와 보여주기

KH2Wone committed:
Add : Mock Data 작업, 댓글 기능 오류 작업중

KH2Wone committed:
[Fix] 댓글 입력이 안되는 버그 해결 및 피드 분리
- Mock Data 연결 후 댓글 기능 안되는 버그를 발견해 해결
- map으로 뿌려줄 각각의 Feed와 Feed들이 그려질 Feeds 파일 생성(분리)
